Summary char limit - fixes #1075

becky-gilbert opened the pull request:
As discussed in #1075, this PR is to reduce the text input limit for study preview summaries from 500 to 300 characters.

Here is what the new limit looks like (test study):

<img width="1187" alt="Screenshot 2023-02-14 at 1 56 07 PM" src="https://user-images.githubusercontent.com/9041788/218875752-dc3b47a8-aebe-4f5b-852d-4678bef6bc82.png">

The text input field will now stop accepting characters after 300. The limit is also indicated via the preview summary help text:

<img width="936" alt="Screenshot 2023-02-14 at 2 03 39 PM" src="https://user-images.githubusercontent.com/9041788/218875835-fa46e841-42c7-448e-b8d2-06b37e809e2e.png">

This will not change anything for existing studies that already have a preview summary between 300 and 500 characters. However, if the researcher edits the study and tries to save it, they will get a form validation error and banner message:

<img width="1031" alt="Screenshot 2023-02-14 at 2 21 19 PM" src="https://user-images.githubusercontent.com/9041788/218878119-d2e35f08-1010-4816-8ddb-a901aebed96e.png">
<img width="942" alt="Screenshot 2023-02-14 at 2 21 41 PM" src="https://user-images.githubusercontent.com/9041788/218878125-d8547e86-109a-492a-af3f-81e7fd2544d4.png">

okaycj reviewed (changes requested):
Nice work! If the model is updated, there will usually be a need for a migration.  Would you mind running `makemigration` and include it in this PR?

becky-gilbert commented:
> If the model is updated, there will usually be a need for a migration. Would you mind running makemigration and include it in this PR?

@okaycj Done! But when I ran `make migrate` after producing this migration file, I get the error: "django.db.utils.DataError: value too long for type character varying(300)". Looks like it's not happy with existing preview summaries in the database that are longer than the character limit. Did I need to do anything to fix this?

okaycj commented:
> @okaycj Done! But when I ran `make migrate` after producing this migration file, I get the error: "django.db.utils.DataError: value too long for type character varying(300)". Looks like it's not happy with existing preview summaries in the database that are longer than the character limit. Did I need to do anything to fix this?

If I do the following I have no issues:
1. Remove DB
2. Add DB from staging
3. Run migrations

Here's the output:
```
Projects/lookit-api - [summary-char-limit] » make migrate
docker compose run --rm web poetry run ./manage.py migrate
[+] Running 1/0
 ⠿ Container lookit-api-db  Running                                                                                                                                                                    0.0s
-------------------Why yes, we are using local assets!-------------------------
Operations to perform:
  Apply all migrations: accounts, admin, auth, authtoken, contenttypes, django_celery_beat, guardian, sessions, sites, studies, web
Running migrations:
  Applying studies.0089_auto_20230217_1855... OK
Projects/lookit-api - [summary-char-limit] »
```

becky-gilbert commented:
I've updated this PR to use the approach discussed in #1075. 
The new migration file finds any studies with preview summaries greater than 300 characters, and for each of these:
1. reduces the preview summary length to 297
2. adds "..." to the end of the preview summary to indicate that it has been clipped
3. copies the entire original preview summary text into the start of the study's short description section. 

Here's what this change looks like:

# Before migration

## Example preview summary - 445 characters
![pre_migration_preview](https://user-images.githubusercontent.com/9041788/233206817-a1412470-ce46-4a22-838a-f447f24731da.png)

## Example description
![pre_migration_description](https://user-images.githubusercontent.com/9041788/233206852-e44d684f-3088-4a66-9ae8-c0ce8631a2bf.png)

# After migration 

## Preview summary - 300 characters
![post_migration_preview](https://user-images.githubusercontent.com/9041788/233207818-e74719fc-ff49-47bb-98d1-93e91e249249.png)

## Description - original summary appended to the start
![post_migration_description](https://user-images.githubusercontent.com/9041788/233207832-b957e003-a7ac-4ac0-a69f-463681a3f4c4.png)

The only studies that are affected are those with preview summaries greater than 300 characters.

@mekline just checking that the last 2 screenshots are what you had in mind, and that are you're still happy with 300 characters as the limit?

# Implementation notes:

@okaycj I'm using print statements to get the names and IDs of studies that are affected by the data alteration, so that we can notify researchers and they can check/edit their summaries/descriptions after this change. Let me know if there's a better way to identify the studies that are affected.

I added a reverse function so that we can undo these changes if necessary, but it's not perfect. After identifying study summaries that have been clipped, it finds the original (300+ character) preview summary by grabbing the text from the start of the study description up to the first newline character after the 300th character. The idea behind this method is to locate the end of the original summary text by finding the newline character that was added after the summary when it was copied into the start of the description. But this means that, if the original preview summary happened to contain a newline between the 300th and 500th characters, then the reverse function will split the description text in the wrong place. I think this scenario isn't super likely to occur, plus we'll have a list of the studies that are affected by this reversal, so we can notify researchers and have them check their summary and description text.
However, maybe we'd rather not have a reversal function than one that doesn't work in all scenarios - let me know if that's the case and I can remove it.

mekline reviewed (approved):
Correctly interpreted @becky, thank you!  Let's remember to announce this change to the userbase when it goes into production

okaycj commented:
> @okaycj I'm using print statements to get the names and IDs of studies that are affected by the data alteration, so that we can notify researchers and they can check/edit their summaries/descriptions after this change. Let me know if there's a better way to identify the studies that are affected.

That is very much a reasonable way to do this, except how are you going to get those values from production?

becky-gilbert commented:
> That is very much a reasonable way to do this, except how are you going to get those values from production?

@okaycj I was hoping the print statements would show up in the GCP cloud console under Kubernetes -> Workloads -> production-migration -> Logs. In that area I can see some logs about the last migration on production (e.g. "Applying studies.0089_alter_lab_slug... OK"). But I realize the `print` out location might depend on the environment settings, and I might just be seeing it in the same place as the other migration logs when things are running in local/debug mode. Do you know for sure? If not, maybe we can try it on staging?
